File: internal/types/contract_review.go
package types

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

type ContractReviewStatus string
type ContractReviewRiskLevel string
type ContractReviewParty string
type ContractReviewBulkAction string
type ContractReviewQualityStatus string
type ContractReviewEvidenceRole string
type ContractReviewFactStatus string

const (
	ContractReviewStatusDraft            ContractReviewStatus = "draft"
	ContractReviewStatusUploading        ContractReviewStatus = "uploading"
	ContractReviewStatusReady            ContractReviewStatus = "ready"
	ContractReviewStatusAnalyzing        ContractReviewStatus = "analyzing"
	ContractReviewStatusReviewingClauses ContractReviewStatus = "reviewing_clauses"
	ContractReviewStatusCompleted        ContractReviewStatus = "completed"
	ContractReviewStatusFailed           ContractReviewStatus = "failed"
	ContractReviewStatusCancelled        ContractReviewStatus = "cancelled"

	ContractReviewRiskHigh   ContractReviewRiskLevel = "high"
	ContractReviewRiskMedium ContractReviewRiskLevel = "medium"
	ContractReviewRiskLow    ContractReviewRiskLevel = "low"

	ContractReviewPartyCustomer ContractReviewParty = "customer"
	ContractReviewPartyVendor   ContractReviewParty = "vendor"
	ContractReviewPartyNeutral  ContractReviewParty = "neutral"

	ContractReviewQualityPending  ContractReviewQualityStatus = "pending"
	ContractReviewQualityValid    ContractReviewQualityStatus = "valid"
	ContractReviewQualityDegraded ContractReviewQualityStatus = "degraded"
	ContractReviewQualityInvalid  ContractReviewQualityStatus = "invalid"
	ContractReviewQualityStale    ContractReviewQualityStatus = "stale"
	ContractReviewQualityLegacy   ContractReviewQualityStatus = "legacy"

	ContractReviewEvidencePrimary       ContractReviewEvidenceRole = "primary"
	ContractReviewEvidenceSupporting    ContractReviewEvidenceRole = "supporting"
	ContractReviewEvidenceContradicting ContractReviewEvidenceRole = "contradicting"

	ContractReviewFactPresent   ContractReviewFactStatus = "present"
	ContractReviewFactAmbiguous ContractReviewFactStatus = "ambiguous"
	ContractReviewFactConflict  ContractReviewFactStatus = "conflict"
	ContractReviewFactNotFound  ContractReviewFactStatus = "not_found"

	ContractReviewBulkArchive ContractReviewBulkAction = "archive"
	ContractReviewBulkRestore ContractReviewBulkAction = "restore"
	ContractReviewBulkDelete  ContractReviewBulkAction = "delete"
)

type ContractReviewBulkItem struct {
	ID      string `json:"id"`
	Success bool   `json:"success"`
	Error   string `json:"error,omitempty"`
}

type ContractReviewBulkResult struct {
	Action    ContractReviewBulkAction `json:"action"`
	Requested int                      `json:"requested"`
	Succeeded int                      `json:"succeeded"`
	Failed    int                      `json:"failed"`
	Items     []ContractReviewBulkItem `json:"items"`
}

type ContractReview struct {
	ID               string               `json:"id" gorm:"type:varchar(36);primaryKey"`
	TenantID         uint64               `json:"tenant_id" gorm:"not null;index"`
	UserID           string               `json:"user_id" gorm:"type:varchar(64);not null;index"`
	Title            string               `json:"title" gorm:"type:varchar(512);not null"`
	TitleCustomized  bool                 `json:"title_customized" gorm:"not null;default:false"`
	Status           ContractReviewStatus `json:"status" gorm:"type:varchar(32);not null;index"`
	Progress         int                  `json:"progress" gorm:"not null;default:0"`
	PlaybookID       string               `json:"playbook_id" gorm:"type:varchar(64);not null"`
	PlaybookVersion  string               `json:"playbook_version" gorm:"type:varchar(32);not null"`
	RepresentedParty ContractReviewParty  `json:"represented_party" gorm:"type:varchar(16);not null"`
	ModelID          string               `json:"model_id" gorm:"type:varchar(64);not null;default:''"`
	AnalysisRunID    string               `json:"analysis_run_id,omitempty" gorm:"type:varchar(64);not null;default:'';index"`
	ConfigHash       string               `json:"-" gorm:"type:varchar(64);not null;default:''"`
	SourceRevision   string               `json:"source_revision,omitempty" gorm:"type:varchar(128);not null;default:''"`
	SourceTextHash   string               `json:"source_text_hash,omitempty" gorm:"type:varchar(64);not null;default:''"`
	// SourceHash is kept as an internal compatibility alias for deployments
	// that already applied the first quality migration. New API consumers use
	// source_text_hash.
	SourceHash       string                      `json:"-" gorm:"type:varchar(64);not null;default:''"`
	Locator          JSON                        `json:"locator,omitempty" gorm:"type:jsonb;not null;default:'{}'"`
	QualityStatus    ContractReviewQualityStatus `json:"quality_status,omitempty" gorm:"type:varchar(16);not null;default:'pending'"`
	Warnings         JSON                        `json:"warnings,omitempty" gorm:"type:jsonb;not null;default:'[]'"`
	ResourceRef      string                      `json:"-" gorm:"type:text"`
	FileName         string                      `json:"file_name" gorm:"type:varchar(1024);not null;default:''"`
	FileType         string                      `json:"file_type" gorm:"type:varchar(16);not null;default:''"`
	MimeType         string                      `json:"mime_type" gorm:"type:varchar(255);not null;default:''"`
	FileSize         int64                       `json:"file_size" gorm:"not null;default:0"`
	ExtractedContent string                      `json:"-" gorm:"type:text"`
	Metadata         JSON                        `json:"metadata" gorm:"type:jsonb;not null;default:'{}'"`
	Overview         JSON                        `json:"overview" gorm:"type:jsonb;not null;default:'{}'"`
	ErrorMessage     string                      `json:"error_message,omitempty" gorm:"type:text"`
	ArchivedAt       *time.Time                  `json:"archived_at,omitempty" gorm:"index"`
	StartedAt        *time.Time                  `json:"started_at,omitempty"`
	CompletedAt      *time.Time                  `json:"completed_at,omitempty"`
	CreatedAt        time.Time                   `json:"created_at"`
	UpdatedAt        time.Time                   `json:"updated_at"`
	DeletedAt        gorm.DeletedAt              `json:"-" gorm:"index"`
	Clauses          []*ContractReviewClause     `json:"clauses,omitempty" gorm:"foreignKey:ReviewID"`
	Issues           []*ContractReviewIssue      `json:"issues,omitempty" gorm:"foreignKey:ReviewID"`
}

func (r *ContractReview) BeforeCreate(_ *gorm.DB) error {
	if r.ID == "" {
		r.ID = uuid.NewString()
	}
	if r.Title == "" {
		r.Title = "Untitled Review"
	}
	if r.Status == "" {
		r.Status = ContractReviewStatusDraft
	}
	if r.PlaybookID == "" {
		r.PlaybookID = "general-contract-review"
	}
	if r.PlaybookVersion == "" {
		r.PlaybookVersion = "1.0"
	}
	if r.RepresentedParty == "" {
		r.RepresentedParty = ContractReviewPartyNeutral
	}
	if r.QualityStatus == "" {
		r.QualityStatus = ContractReviewQualityPending
	}
	if len(r.Metadata) == 0 {
		r.Metadata = JSON(`{}`)
	}
	if len(r.Overview) == 0 {
		r.Overview = JSON(`{}`)
	}
	if len(r.Locator) == 0 {
		r.Locator = JSON(`{}`)
	}
	if len(r.Warnings) == 0 {
		r.Warnings = JSON(`[]`)
	}
	return nil
}

type ContractReviewClause struct {
	ID             string    `json:"id" gorm:"type:varchar(36);primaryKey"`
	ReviewID       string    `json:"review_id" gorm:"type:varchar(36);not null;index"`
	Sequence       int       `json:"sequence" gorm:"not null"`
	Title          string    `json:"title" gorm:"type:varchar(512);not null"`
	Excerpt        string    `json:"excerpt" gorm:"type:text"`
	SourceStart    int       `json:"source_start" gorm:"not null;default:0"`
	SourceEnd      int       `json:"source_end" gorm:"not null;default:0"`
	EvidenceID     string    `json:"evidence_id,omitempty" gorm:"type:varchar(128);not null;default:''"`
	SourceRevision string    `json:"source_revision,omitempty" gorm:"-"`
	ReviewStatus   string    `json:"review_status" gorm:"type:varchar(16);not null;default:'pending'"`
	IssueCount     int       `json:"issue_count" gorm:"not null;default:0"`
	CreatedAt      time.Time `json:"created_at"`
	UpdatedAt      time.Time `json:"updated_at"`
}

func (c *ContractReviewClause) BeforeCreate(_ *gorm.DB) error {
	if c.ID == "" {
		c.ID = uuid.NewString()
	}
	if c.ReviewStatus == "" {
		c.ReviewStatus = "pending"
	}
	return nil
}

type ContractReviewEvidence struct {
	EvidenceID     string                     `json:"evidence_id"`
	Role           ContractReviewEvidenceRole `json:"role,omitempty"`
	SourceRevision string                     `json:"source_revision,omitempty"`
	SourceTextHash string                     `json:"source_text_hash,omitempty"`
	SourceStart    int                        `json:"source_start,omitempty"`
	SourceEnd      int                        `json:"source_end,omitempty"`
	PageStart      int                        `json:"page_start,omitempty"`
	PageEnd        int                        `json:"page_end,omitempty"`
	Status         string                     `json:"status,omitempty"`
}

type ContractReviewIssue struct {
	ID             string                   `json:"id" gorm:"type:varchar(36);primaryKey"`
	ReviewID       string                   `json:"review_id" gorm:"type:varchar(36);not null;index"`
	ClauseID       string                   `json:"clause_id" gorm:"type:varchar(36);not null;index"`
	Fingerprint    string                   `json:"-" gorm:"type:varchar(64);not null;uniqueIndex"`
	Sequence       int                      `json:"sequence" gorm:"not null"`
	RiskLevel      ContractReviewRiskLevel  `json:"risk_level" gorm:"type:varchar(16);not null"`
	Category       string                   `json:"category" gorm:"type:varchar(80);not null;default:''"`
	FindingType    string                   `json:"finding_type" gorm:"type:varchar(80);not null;default:''"`
	Title          string                   `json:"title" gorm:"type:varchar(512);not null"`
	Explanation    string                   `json:"explanation" gorm:"type:text;not null"`
	OriginalQuote  string                   `json:"original_quote" gorm:"type:text;not null"`
	Suggestion     string                   `json:"suggestion" gorm:"type:text;not null"`
	EvidenceRefs   JSON                     `json:"evidence_refs,omitempty" gorm:"type:jsonb;not null;default:'[]'"`
	Evidence       []ContractReviewEvidence `json:"evidence,omitempty" gorm:"-"`
	EvidenceStatus string                   `json:"evidence_status,omitempty" gorm:"-"`
	SourceRevision string                   `json:"source_revision,omitempty" gorm:"-"`
	SourceStart    int                      `json:"source_start" gorm:"not null;default:0"`
	SourceEnd      int                      `json:"source_end" gorm:"not null;default:0"`
	CreatedAt      time.Time                `json:"created_at"`
	UpdatedAt      time.Time                `json:"updated_at"`
}

func (i *ContractReviewIssue) BeforeCreate(_ *gorm.DB) error {
	if i.ID == "" {
		i.ID = uuid.NewString()
	}
	return nil
}

type ContractReviewPlaybook struct {
	ID          string `json:"id"`
	Name        string `json:"name"`
	Description string `json:"description"`
	Version     string `json:"version"`
}

type ContractReviewTaskPayload struct {
	TenantID      uint64 `json:"tenant_id"`
	UserID        string `json:"user_id"`
	ReviewID      string `json:"review_id"`
	AnalysisRunID string `json:"analysis_run_id,omitempty"`
	ConfigHash    string `json:"config_hash,omitempty"`
}

// ContractReviewFact is a normalized, evidence-backed fact candidate. Facts
// are stored in the compatible overview JSON for the MVP; keeping this type in
// the domain package gives later normalized fact tables the same wire shape.
type ContractReviewFact struct {
	Type            string                   `json:"type"`
	Key             string                   `json:"key"`
	Value           string                   `json:"value"`
	NormalizedValue string                   `json:"normalized_value,omitempty"`
	Unit            string                   `json:"unit,omitempty"`
	Currency        string                   `json:"currency,omitempty"`
	Condition       string                   `json:"condition,omitempty"`
	Status          ContractReviewFactStatus `json:"status"`
	EvidenceQuote   string                   `json:"evidence_quote,omitempty"`
	SourceStart     int                      `json:"source_start,omitempty"`
	SourceEnd       int                      `json:"source_end,omitempty"`
	EvidenceRefs    []string                 `json:"evidence_refs,omitempty"`
}

type ContractReviewWarning struct {
	Code        string `json:"code"`
	Message     string `json:"message"`
	ClauseID    string `json:"clause_id,omitempty"`
	EvidenceID  string `json:"evidence_id,omitempty"`
	SourceStart int    `json:"source_start,omitempty"`
	SourceEnd   int    `json:"source_end,omitempty"`
}

File: frontend/src/api/contract-review.ts
import { fetchEventSource } from '@microsoft/fetch-event-source'

import { getApiBaseUrl } from '@/utils/api-base'
import { del, get, patch, post, postUpload } from '@/utils/request'

export type ReviewStatus = 'draft' | 'uploading' | 'ready' | 'analyzing' | 'reviewing_clauses' | 'completed' | 'failed' | 'cancelled'
export type RiskLevel = 'high' | 'medium' | 'low'
export type RepresentedParty = 'customer' | 'vendor' | 'neutral'
export type EvidenceStatus = 'pending' | 'located' | 'legacy_exact' | 'multiple_matches' | 'not_found' | 'version_mismatch' | 'unsupported' | 'error'
export type LocatorLoadStatus = 'idle' | 'loading' | 'ready' | 'unsupported' | 'error'
export type ConfidenceLevel = 'high' | 'medium' | 'low' | 'unknown'
export type ReviewQualityStatus = 'pending' | 'valid' | 'degraded' | 'invalid' | 'stale' | 'legacy'

// These are intentionally open string unions. A playbook may add a category
// or finding type before the UI ships a new translation; the fallback label
// keeps the result readable without rejecting an otherwise valid response.
export type ReviewCategory = string
export type ReviewFindingType = string

export interface ReviewEvidence {
  source_start?: number
  source_end?: number
  source_revision?: string
  source_text_hash?: string
  unit_ids?: string[]
  page_start?: number
  page_end?: number
  status?: EvidenceStatus
}

export interface ReviewFact {
  type?: string
  key?: string
  label?: string
  value?: string
  text?: string
  normalized_value?: string
  unit?: string
  currency?: string
  condition?: string
  status?: string
  evidence_quote?: string
  evidence_refs?: string[]
  source_start?: number
  source_end?: number
}

export type ReviewFactValue = string | ReviewFact

export interface ReviewWarning {
  code?: string
  message?: string
  clause_id?: string
  evidence_id?: string
}

export type ReviewWarningValue = string | ReviewWarning

export interface ContractReviewLocatorUnit {
  unit_id: string
  kind?: string
  parent_id?: string
  page?: number
  source_start: number
  source_end: number
  text?: string
  // Optional rendered offsets let the viewer disambiguate repeated text
  // without guessing. They are normalized document offsets, end-exclusive.
  rendered_start?: number
  rendered_end?: number
}

export interface ContractReviewLocator {
  version?: number
  offset_unit?: string
  source_hash?: string
  source_length?: number
  review_id?: string
  source_revision?: string
  source_text_hash?: string
  // `text` is the canonical extracted source text in newer responses.
  // `source_text` is accepted as an explicit alias for compatibility.
  text?: string
  source_text?: string
  units?: ContractReviewLocatorUnit[]
  source_units_json?: ContractReviewLocatorUnit[] | string
}

export interface ReviewClause {
  id: string
  review_id: string
  sequence: number
  title: string
  excerpt: string
  source_start: number
  source_end: number
  source_revision?: string
  review_status: string
  issue_count: number
}

export interface ReviewIssue {
  id: string
  review_id: string
  clause_id: string
  sequence: number
  risk_level: RiskLevel
  title: string
  explanation: string
  original_quote: string
  suggestion: string
  source_start: number
  source_end: number
  source_revision?: string
  evidence?: ReviewEvidence | ReviewEvidence[]
  category?: ReviewCategory
  finding_type?: ReviewFindingType
  evidence_refs?: string[]
  evidence_status?: EvidenceStatus
  confidence?: ConfidenceLevel
  confidence_score?: number
}

export interface ReviewOverview {
  overall_risk?: RiskLevel
  executive_summary?: string
  contract_type?: string
  parties?: string[]
  facts?: ReviewFactValue[]
  quality_status?: ReviewQualityStatus
  warnings?: ReviewWarningValue[]
  key_recommendations?: string[]
  risk_counts?: Record<RiskLevel, number>
}

export interface ReviewPlaybook { id: string; name: string; description: string; version: string }
export type ContractReviewBulkAction = 'archive' | 'restore' | 'delete'
export interface ContractReviewBulkItem { id: string; success: boolean; error?: string }
export interface ContractReviewBulkResult {
  action: ContractReviewBulkAction
  requested: number
  succeeded: number
  failed: number
  items: ContractReviewBulkItem[]
}

export interface ContractReview {
  id: string
  title: string
  title_customized: boolean
  status: ReviewStatus
  progress: number
  playbook_id: string
  playbook_version: string
  represented_party: RepresentedParty
  model_id: string
  file_name: string
  file_type: '.pdf' | '.docx' | ''
  mime_type: string
  file_size: number
  metadata: Record<string, string>
  overview: ReviewOverview
  source_revision?: string
  source_hash?: string
  source_text_hash?: string
  locator?: ContractReviewLocator
  facts?: ReviewFactValue[]
  quality_status?: ReviewQualityStatus
  warnings?: ReviewWarningValue[]
  error_message?: string
  archived_at?: string
  started_at?: string
  completed_at?: string
  created_at: string
  updated_at: string
  clauses?: ReviewClause[]
  issues?: ReviewIssue[]
}

export interface ApiResponse<T> { success: boolean; data: T }

export const listContractReviews = (archived = false) => get<ApiResponse<ContractReview[]>>(`/api/v1/contract-reviews?archived=${archived}`)
export const listContractReviewPlaybooks = () => get<ApiResponse<ReviewPlaybook[]>>('/api/v1/contract-review-playbooks')
export const createContractReview = () => post<ApiResponse<ContractReview>>('/api/v1/contract-reviews')
export const getContractReview = (id: string) => get<ApiResponse<ContractReview>>(`/api/v1/contract-reviews/${id}`)
export const updateContractReview = (id: string, data: Partial<Pick<ContractReview, 'title' | 'playbook_id' | 'represented_party' | 'model_id'>> & { archived?: boolean }) => patch<ApiResponse<ContractReview>>(`/api/v1/contract-reviews/${id}`, data)
export const deleteContractReview = (id: string) => del(`/api/v1/contract-reviews/${id}`)
export const bulkContractReviews = (action: ContractReviewBulkAction, ids: string[]) =>
  post<ApiResponse<ContractReviewBulkResult>>(`/api/v1/contract-reviews/bulk/${action}`, { ids })
export const startContractReview = (id: string) => post<ApiResponse<ContractReview>>(`/api/v1/contract-reviews/${id}/start`)
export const retryContractReview = (id: string) => post<ApiResponse<ContractReview>>(`/api/v1/contract-reviews/${id}/retry`)
export const cancelContractReview = (id: string) => post<ApiResponse<ContractReview>>(`/api/v1/contract-reviews/${id}/cancel`)
export const getContractReviewDocument = (id: string) => get<ArrayBuffer>(`/api/v1/contract-reviews/${id}/document/preview`, { responseType: 'arraybuffer' })
export const getContractReviewLocator = (id: string) =>
  get<ApiResponse<ContractReviewLocator>>(`/api/v1/contract-reviews/${id}/document/locator`)

export const uploadContractReviewDocument = (id: string, file: File, onProgress?: (value: number) => void) => {
  const data = new FormData()
  data.append('file', file)
  return postUpload(`/api/v1/contract-reviews/${id}/document`, data, (event) => {
    if (event.total) onProgress?.(Math.round((event.loaded / event.total) * 100))
  }) as Promise<ApiResponse<ContractReview>>
}

export function streamContractReview(id: string, signal: AbortSignal, onSnapshot: (review: ContractReview) => void) {
  const token = localStorage.getItem('weknora_token') || ''
  const tenant = localStorage.getItem('weknora_selected_tenant_id') || ''
  return fetchEventSource(`${getApiBaseUrl()}/api/v1/contract-reviews/${id}/events`, {
    method: 'GET', signal, openWhenHidden: true,
    headers: {
      Authorization: `Bearer ${token}`,
      'Accept-Language': localStorage.getItem('locale') || 'zh-CN',
      ...(tenant ? { 'X-Tenant-ID': tenant } : {}),
    },
    async onopen(response) { if (!response.ok) throw new Error(`HTTP ${response.status}`) },
    onmessage(event) {
      if (event.event !== 'snapshot' || !event.data) return
      try { onSnapshot(JSON.parse(event.data) as ContractReview) } catch { /* Ignore malformed frames; snapshot polling remains authoritative. */ }
    },
    onerror() { if (!signal.aborted) return 1500 },
  })
}
